Add local cache for noderesourcetopology using the reserve plugin [2/3] - implementation

fromanirh opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
Implement the noderesourcetopology pessimistic cache allocation, building on top of #315 .
Includes very basic integration tests. More elaborate integration tests, and most notably e2e tests, will require a significant investment in infrastructure (both code and machines), so postponing for now until we have more clarity.

#### Which issue(s) this PR fixes:
Fixes N/A

#### Special notes for your reviewer:
design doc and release note added on #315 

The work is partitioned as follows:
- the initial PR #315 was expanded in scope as flaws in the original old design emerged during the review. We now use #315 to add documentation, context and the very basic groundwork.
- the followup #437 extends and completes #315. PRs are kept split to make the review easier (or to attempt so) and to preserve original context of #315 and all the discussion. However, we need both #315 and #437 (the latter bundles the former) to have the bare minimum required functionality
- finally, we have #439 which extends further #437. Changes added here aren't part of the minimal core but add features needed in real-world scenarios. This work can be reviewed separately and builds on top of #437. This completes the feature.

fromanirh commented:
/release-note-none

fromanirh commented:
depends on and includes #315

swatisehgal reviewed (commented):
Took a pass at the implementation. 
It would be great if we can split the implementation commit further for ease of review? The obvious one to me is to split the indexer workaround logic into a separate commit which will make for an easier update in case we move 1.25 while this PR is being reviewed.

Will do another round to review to cover the integration tests.

fromanirh commented on pkg/noderesourcetopology/cache/indexer.go:
```diff
@@ -0,0 +1,173 @@
+)
+
+/*
+ * This is needed because of https://github.com/kubernetes/kubernetes/issues/110660
+ * Fixed in kubernetes 1.25.0, but enhancements are not backported.
+ * So we need to wait until the scheduler plugins repo consumes k8s >= 1.25.0, then
+ * We can just add a client-go/cache Indexer and get rid of this workaround:
```
fair point. The indexer was born out of necessity but the concept kinda grew on me. We may want to at least discuss if we want to keep it around in the long run. 
That said, the challenge here is that the overallocating cache doesn't work with either our home-grown indexer or without using the `client-go` native solution. It would be great to always have the tree in a working state after each commit. I'll think about possible approaches here.

fromanirh replied:
I think I found a good approach which also trims down this pretty big commit. I'll push updates to this PR and only after fix #315

fromanirh commented on pkg/noderesourcetopology/cache/overallocating.go:
```diff
@@ -0,0 +1,221 @@
+// state of a dirty node, for all the dirty nodes. If this is the case, the cache of a node can be Flush()ed.
+func (ov *Overallocating) Resync() {
+	// we are not working with a specific pod, so we need a unique key to track this flow
+	logKey := shortuuid.New()
```
a unique identifier constantly used in all the log lines belonging to the same resync iteration helps when we have big logs, or logs with high traffic. By grepping `$logKey`, we know we catch all the relevant log entries. In the same conditions we can't have the same guarantees just using timestamps, so a explicit tracking identifier helps.

swatisehgal replied:
Ack, Thanks. I understand the use of `logkey` now but I was suggesting to have a `logkey` such that we have the ability to determine the previous and the next iteration for a `logkey` to gain a better understanding of why the current resync loop is behaving the way it is. I am not caught up on the use of timestamp but was thinking about something that allows us to get an indication of the order of resync loops. I guess, reviewing the `logkey` in conjunction with timestamp should allow us to achieve that but I was wondering if we could do something better as part of the `logkey` itself.

fromanirh replied:
Sorry, I misunderstood you previous comment and gave only a partially relevent answer. I'm inclined to rely to the timestamp any sane logging format (and scheduler plugin's default) should provide to add the timestamp information. If timestamp is a given, adding it again in the `logKey` feels redundant (but not wrong). Anyway I'll revisit this assumption; if nothing else, we can possibly shred dependencies.

fromanirh replied:
done, let's discuss in the review of the next upload.

fromanirh replied:
it now looks like
```
I1024 13:31:05.915942       1 overallocating.go:196] "nrtcache: resync NodeTopology cache starting" logKey="resync1666618265915"
I1024 13:31:05.915997       1 overallocating.go:234] "nrtcache: resync NodeTopology cache complete" logKey="resync1666618265915"
```

fromanirh commented on pkg/noderesourcetopology/cache/overallocating.go:
```diff
@@ -0,0 +1,221 @@
+
+		pfpExpected := podFingerprintForNodeTopology(nrtCandidate)
+		if pfpExpected == "" {
+			klog.V(3).InfoS("nrtcache: missing NodeTopology podset fingerprint data", "logKey", logKey, "node", nodeName)
```
that's true. I'm not sure how can we handle in a more explicit way though. Missing the fingerprint from the node cause that node to be considered full prematurely (because pessimistic allocation) but doesn't impact others.

fromanirh commented on pkg/noderesourcetopology/cache/overallocating.go:
```diff
@@ -0,0 +1,221 @@
+	return nrt
+}
+
+func (ov *Overallocating) MarkNodeDiscarded(nodeName string, pod *corev1.Pod) {
```
Both are done in the parent commit: https://github.com/kubernetes-sigs/scheduler-plugins/pull/437/commits/fb053f39e63aa8966aa7130bc305dd623f65f356

swatisehgal replied:
Ah, thanks. Should have taken a deeper look.

fromanirh replied:
no problem! rearranged the code to make it more obvious (I think)

fromanirh commented on pkg/noderesourcetopology/cache/indexer.go:
```diff
@@ -0,0 +1,173 @@
+	// maps can leak. We can't estimate how frequent or how bad this can be yet.
+	// if this becomes a real bug, the fix is to recreate the inner maps with only the live keys, dropping
+	// the old leaky map. We should probably do this here, on delete, but not every time but once the pod
+	// churn (easily trackavle with a integer counting the deletes) crosses a threshold, whose value is TBD.
```
thanks, fixed

fromanirh commented on pkg/noderesourcetopology/cache/store.go:
```diff
@@ -0,0 +1,218 @@
+
+// nrtStore maps the NRT data by node name. It is not thread safe and needs to be protected by a lock.
+// data is intentionally copied each time it enters and exists the store. E.g, no pointer sharing.
+type nrtStore map[string]*topologyv1alpha1.NodeResourceTopology
```
Uhm, this could require a redesign of the internal objects. I'll put some thought on it.

fromanirh replied:
amended the code comments, most notably on `resync` to explain the reasoning. Let's discuss in the next upload.

swatisehgal reviewed (commented):
The implementation looks good to me! I will review the unit tests is more detail in my subsequent review.

Great start on the integration tests as well. I was wondering if can go a bit further to highlight the reserve plugin cache implementation and check the correctness of updates made to it? Maybe we add a test where a pod is scheduled by Topology-aware scheduler and reserve plugin updates the cache. Can we check if the cache was updated correctly? Does the actual state of the cache match the expected state?

fromanirh commented on pkg/noderesourcetopology/cache/overallocating.go:
```diff
@@ -0,0 +1,234 @@
+// too aggressive resync attempts, so to more, likely unnecessary, computation work on the scheduler side.
+func (ov *Overallocating) Resync() {
+	// we are not working with a specific pod, so we need a unique key to track this flow
+	logKey := logKeyFromTime()
```
thanks!

Tal-or commented on pkg/noderesourcetopology/cache/store.go:
```diff
@@ -0,0 +1,218 @@
+}
+
+// resourceCountersFromPod computes the cumulative resource requests required by the given pod.
+func resourceCountersFromPod(pod *corev1.Pod) corev1.ResourceList {
```
Maybe use https://github.com/kubernetes-sigs/scheduler-plugins/blob/master/pkg/util/resource.go#L50 instead which covers init containers as well

fromanirh replied:
thanks this seems better indeed!

Tal-or reviewed (commented):
Looks very impressive. 
I left a couple of comments though, but nothing serious.

fromanirh commented on pkg/noderesourcetopology/cache/overallocating.go:
```diff
@@ -0,0 +1,234 @@
+	for _, nodeName := range nodeNames {
+		nrtCandidate, err := ov.lister.Get(nodeName)
+		if err != nil || nrtCandidate == nil {
+			klog.V(3).InfoS("nrtcache: missing NodeTopology", "logKey", logKey, "node", nodeName)
```
yes, this won't hurt. Thanks!

fromanirh commented on pkg/noderesourcetopology/cache/store.go:
```diff
@@ -0,0 +1,218 @@
+	return obj.DeepCopy()
+}
+
+// Upddate adds or replace the Node Resource Topology associated to a node. Always do a copy.
```
thanks, fixed.

fromanirh commented on pkg/noderesourcetopology/cache/overallocating.go:
```diff
@@ -0,0 +1,234 @@
+// DirtyNodeNames returns a slice of all the node names which have been discarded previously,
+// so which are supposed to be `dirty` in the cache.
+// A node can be discarded for two reasons:
+// 1. it legititmately cannot fit containers because it has not enough free resources
```
thanks, fixed

fromanirh commented on pkg/noderesourcetopology/cache/indexer.go:
```diff
@@ -0,0 +1,173 @@
+	logKey := nn.String()
+
+	if pod.Spec.NodeName == "" {
+		klog.V(4).InfoS("nrtcache: nni WRN", "logKey", logKey, "node", "EMPTY", "podUID", pod.UID, "phase", pod.Status.Phase)
```
sure, renamed to `WARN`

fromanirh commented on pkg/noderesourcetopology/cache/indexer.go:
```diff
@@ -0,0 +1,173 @@
+				klog.V(3).InfoS("nrtcache: nni: update unsupported object %T", newObj)
+				return
+			}
+			nni.addPod(pod)
```
You're right. This was an initial attempt backtracked and never fully removed up until now. Let's get rid of half-done code.

fromanirh commented on test/integration/noderesourcetopology_cache_test.go:
```diff
@@ -0,0 +1,441 @@
+
+	for _, tt := range []testCase{
+		{
+			name: "Filtering out nodes that cannot fit resources on a single numa node in case of Guaranteed pod",
```
yes, good point. Adding them.

fromanirh commented:
trivial rebase. Testcases (https://github.com/kubernetes-sigs/scheduler-plugins/pull/437#discussion_r1006888245) still pending.

fromanirh commented:
I'd like to mention that we are building a [extensive and growing suite of e2e tests](https://github.com/openshift-kni/numaresources-operator/blob/main/test/e2e/serial/README.md) which cover the behavior of this plugin.
We'd love to contribute back and add e2e tests here alongside the plugin. We have two problems though:

1. we need to have enough machines (2+) powerful enough (each with 2+ NUMA nodes) in CI to be able to perform meaningful tests. RH runs the aforementioned suite regularly on internal platforms we can't unfortunately give access to.
2. we need to orchestrate the deployment of the NUMA-aware scheduling stack. We have options, this is doable but will require quite some work to get there.

We will focus on integration test as possible, but the existing integration tests infra is a partial fit to exercise this code.

fromanirh commented:
The PR is reviewable and the first two commits (indexer utility, implementation proper) are in the current final form, but the last commit (integration test) is still WIP as I work to expand/improve the coverage in the context of the current integration test framework.

fromanirh commented:
addressed the comments about the integration tests.

fromanirh commented:
I've been trying to add integration test covering the resync loop. I can't find yet a good way to update NRT objects and wait for the resync to happen. Do we have integration tests doing something similar in the codebase so I can have some examples?

swatisehgal commented on pkg/noderesourcetopology/cache/overallocating_test.go:
```diff
@@ -0,0 +1,792 @@
+	nrtCache.UnreserveNodeResources("node1", testPod)
+
+	nrtObj := nrtCache.GetByNode("node1", testPod)
+	for _, zone := range nrtObj.Zones {
+		for _, zoneRes := range zone.Resources {
+			switch zoneRes.Name {
+			case string(corev1.ResourceCPU):
+				if zoneRes.Available.Cmp(resource.MustParse("30")) != 0 {
+					t.Errorf("quantity mismatch in zone %q", zoneRes.Name)
+				}
+			case string(corev1.ResourceMemory):
+				if zoneRes.Available.Cmp(resource.MustParse("60Gi")) != 0 {
+					t.Errorf("quantity mismatch in zone %q", zoneRes.Name)
+				}
+			}
+		}
+	}
+}
```
I think we should streamline this logic as the NRT object has not changed:
```
if !reflect.DeepEqual(nrtObj, nodeTopologies[0]) {
		t.Fatalf("unexpected object from cache\ngot: %s\nexpected: %s\n", dumpNRT(nrtObj), dumpNRT(nodeTopologies[0]))
	}
```

fromanirh replied:
right, done.

swatisehgal reviewed (commented):
Did a partial review, will continue with it on Monday.

Huang-Wei commented on pkg/noderesourcetopology/cache/overallocating.go:
```diff
@@ -0,0 +1,238 @@
+
+		klog.V(6).InfoS("nrtcache: trying to resync NodeTopology", "logKey", logKey, "node", nodeName, "fingerprint", pfpExpected)
+
+		err = checkPodFingerprintForNode(logKey, ov.indexer, nodeName, pfpExpected)
```
I guess this is always consistent as ov.indexer is always consistent with ov.lister, and hence consistent with pfpExpected. We should compare the reserved pods (instead of pods in ov.indexer) with pfpExpected?

fromanirh replied:
Uhm, if we stop caching NRT data (per suggestion in https://github.com/kubernetes-sigs/scheduler-plugins/pull/437#discussion_r1021394760) we can indeed compare the pods alone

Huang-Wei reviewed (commented):
I'm not sure I fully understanding the logic here. In particular:

- where to store individual pods are (over-)reserved? (`assumedResources` doesn't track pods)
- how to determine if the latest NRT is good to be synced so as to invalidate the (over-)reserved pods

fromanirh commented:
> I'm not sure I fully understand the logic here. In particular:
> 
>     * where to store the pods that are (over-)reserved? (`assumedResources` doesn't track pods)

The current implementation does not store over-reserved pods, but it takes the resource request (the exclusively-allocatable ones) for these pods and adjust the NRT availability based on that.
I choose this path because the focus here is IMO more on the NRT availability than the pods per se. But there are multiple valid ways to tackle this, so I'm open to change direction.

>     * how to determine if the latest NRT is good to be synced so as to invalidate the (over-)reserved pods

This is completely done in the `Resync` function. In a nutshell, we iterate over suspected nodes, taking the latest NRT data available and checking the expected state with the received state. If they match, we invalidate our existing cache and we trust the latest NRT data.

fromanirh commented on pkg/noderesourcetopology/cache/overallocating.go:
```diff
@@ -0,0 +1,238 @@
+
+	klog.V(3).InfoS("nrtcache: initializing", "objects", len(nrtObjs))
+	return &Overallocating{
+		nrts:             newNrtStore(nrtObjs),
```
This is a interesting take. This design stem by the fact we started with the less intrusive change form the internal API perspective and we kept the NRT data as focal point. But it can totally work the way you are describing, keeping track of pods and invalidating them when the latest NRT matches.

The problem here is to compute the pod fingerprinting, which has to be done at each get attempt. This is potentially costly, 
and this is why we added the resync loop - to throttle and control how many pod fingerprinting we need to do.

I guess we can start with the simplest design and optimize later, then

fromanirh commented on pkg/noderesourcetopology/cache/overallocating.go:
```diff
@@ -0,0 +1,238 @@
+	PodNamespacedNamesByNode(logKey, nodeName string) ([]types.NamespacedName, error)
+}
+
+type Overallocating struct {
```
sure thing to both

fromanirh commented on pkg/noderesourcetopology/cache/indexer.go:
```diff
@@ -0,0 +1,166 @@
+
+type NodeNameIndexer struct {
+	rwLock     sync.RWMutex
+	podsByNode map[string]map[types.UID]types.NamespacedName
```
Will rename accordingly, thanks

fromanirh commented on pkg/noderesourcetopology/cache/store.go:
```diff
@@ -0,0 +1,201 @@
+}
+
+// Update adds or replace the Node Resource Topology associated to a node. Always do a copy.
+func (nrs nrtStore) Update(nrt *topologyv1alpha1.NodeResourceTopology) {
```
yeah, that's exactly the problem here :\ I guess I'll wrap the map in a struct and use a pointer receiver

fromanirh replied:
Or I can just use a pointer receiver. Checking.

fromanirh replied:
No obvious solution yet. I see examples of similar types in the scheduler framework (https://github.com/kubernetes/kubernetes/blob/v1.26.0-beta.0/pkg/scheduler/framework/runtime/registry.go#L70) I'll keep an eye and look for options, but both pointer receivers and especially wrapping in structs seems not great alternatives here.

Huang-Wei replied:
yup, it's because map is a pointer underneath, so functionally it can ensure correctness. But for concert type, if we're updating it, using a pointer receiver is a must.

fromanirh replied:
Right. I guess then I'll have to wrap the maps in structs, it seems to me the somehow less clunky approach.

fromanirh commented on pkg/noderesourcetopology/cache/indexer.go:
```diff
@@ -0,0 +1,166 @@
+	podsByNode map[string]map[types.UID]types.NamespacedName
+}
+
+func NewNodeNameIndexer(podInformer k8scache.SharedInformer) *NodeNameIndexer {
```
ack, will do

fromanirh commented on pkg/noderesourcetopology/cache/store.go:
```diff
@@ -0,0 +1,201 @@
+	klog.V(6).InfoS("nrtcache: podset fingerprint", "logKey", logKey, "pods", len(objs))
+
+	pfp := podfingerprint.NewFingerprint(len(objs))
+	for _, obj := range objs {
+		klog.V(6).InfoS("nrtcache: podset fingerprint", "logKey", logKey, "namespacedName", obj.Namespace+"/"+obj.Name)
+		pfp.Add(obj.Namespace, obj.Name)
+	}
```
this is guaranteed by the package itself to keep its invariant promise: https://pkg.go.dev/github.com/k8stopologyawareschedwg/podfingerprint#pkg-overview

fromanirh commented on pkg/noderesourcetopology/cache/indexer.go:
```diff
@@ -0,0 +1,166 @@
+
+	k8scache "k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+
```
ack, will do

fromanirh commented on pkg/noderesourcetopology/cache/indexer.go:
```diff
@@ -0,0 +1,166 @@
+
+// TODO: tests using multiple goroutines (+ race detector)
+
+type NodeNameIndexer struct {
```
Will change accordingly

fromanirh commented on pkg/noderesourcetopology/cache/indexer.go:
```diff
@@ -0,0 +1,166 @@
+	logKey := nn.String()
+
+	if pod.Spec.NodeName == "" {
+		klog.V(4).InfoS("nrtcache: nni WARN", "logKey", logKey, "node", "EMPTY", "podUID", pod.UID, "phase", pod.Status.Phase)
```
thanks, will try this out and change accordingly

fromanirh commented on pkg/noderesourcetopology/cache/overallocating.go:
```diff
@@ -0,0 +1,238 @@
+	lock             sync.Mutex
+	nrts             nrtStore
+	assumedResources map[string]resourceStore // nodeName -> resourceStore
+	nodeDiscarded    counter
```
sure thing. In a nutshell, it counts how many times a node is filtered out. Needed for the `Resync` logic. (will add an amended form of this text in the comment)

fromanirh commented on pkg/noderesourcetopology/cache/overallocating.go:
```diff
@@ -0,0 +1,238 @@
+	nrts             nrtStore
+	assumedResources map[string]resourceStore // nodeName -> resourceStore
+	nodeDiscarded    counter
+	lister           listerv1alpha1.NodeResourceTopologyLister
+	indexer          NodeIndexer
```
ack, will change

fromanirh commented on pkg/noderesourcetopology/cache/overallocating.go:
```diff
@@ -0,0 +1,238 @@
+// too aggressive resync attempts, so to more, likely unnecessary, computation work on the scheduler side.
+func (ov *Overallocating) Resync() {
+	// we are not working with a specific pod, so we need a unique key to track this flow
+	logKey := logKeyFromTime()
```
exactly. I'm using `logKey` consistently all across the code to group a series of related events in the logs. I'm open to better names to reflect this concept :)

fromanirh replied:
renamed as `logID`

fromanirh commented on pkg/noderesourcetopology/cache/overallocating.go:
```diff
@@ -0,0 +1,238 @@
+	}
+
+	if len(nodes) > 0 {
+		klog.V(4).InfoS("nrtcache: found dirty nodes", "logKey", logKey, "nodes", len(nodes))
```
I think we can recompute the logKey here indeed. I'll experiment with that.

fromanirh replied:
actually dropped this log, we can avoid it.

fromanirh commented on pkg/noderesourcetopology/cache/overallocating.go:
```diff
@@ -0,0 +1,238 @@
+	nodeAssumedResources := ov.assumedResources[nodeName]
+
+	klog.V(6).InfoS("nrtcache NRT", "logKey", logKey, "vanilla", stringify.NodeResourceTopologyResources(nrt))
+	nodeAssumedResources.UpdateNRT(logKey, nrt)
```
I was not aware of this `PostFilter`, so I'll move the logic to over-assume in `Reserve`.
The reason we do the computation here is because we keep a pristine copy of NRT data and we track the overreserve separately. When the client code wants to know the expected availability, we merge the two informations in the returned NRT data.

fromanirh replied:
(rewriting the previous answer which was written too hastily). What we do here is:
1. take a deepcopy of the latest NRT data we have (the latest expected we observed, so the one on which the observed set of pods matched the expected set of pods)
2. update **the local deepcopy** with the requests of the pods, creating the overreserve
3. return that data

so the cached copy is not going to be updated here, but only the local copy. This is safe to be done concurrently.
We track the extra resource to overreserve indeed in `Reserve`.

Huang-Wei replied:
ok, I see. Can we rename `GetByNode` to `GetCachedNRTCopy`, and add a comment like (tweak any words if inaccurate):

```
GetCachedNRTCopy retrieves a NRT copy from cache, and then deducts over-reserved resources if necessary.
It will be used as the source of truth across the Pod's scheduling cycle.
```

fromanirh replied:
Sure, done.

fromanirh commented on pkg/noderesourcetopology/cache/overallocating.go:
```diff
@@ -0,0 +1,238 @@
+	"github.com/k8stopologyawareschedwg/podfingerprint"
+)
+
+type NodeIndexer interface {
+	PodNamespacedNamesByNode(logKey, nodeName string) ([]types.NamespacedName, error)
+}
```
done, thanks

fromanirh commented on pkg/noderesourcetopology/cache/overallocating_test.go:
```diff
@@ -0,0 +1,792 @@
+		nrtCache.MarkNodeDiscarded(nodeName, &corev1.Pod{})
+	}
+
+	// assume noe update which unblocks node-4
```
nice catch :) fixed

fromanirh commented on pkg/noderesourcetopology/cache/overallocating_test.go:
```diff
@@ -0,0 +1,792 @@
+	nicResourceName = "vendor.com/nic1"
+)
+
+func TestInitEmptyLister(t *testing.T) {
```
right. Renamed even more aggressively.

fromanirh commented on pkg/noderesourcetopology/cache/overallocating_test.go:
```diff
@@ -0,0 +1,792 @@
+		t.Fatalf("unexpected error creating cache: %v", err)
+	}
+
+	nodeTopologies := []*topologyv1alpha1.NodeResourceTopology{
+		{
+			ObjectMeta:       metav1.ObjectMeta{Name: "node1"},
+			TopologyPolicies: []string{string(topologyv1alpha1.SingleNUMANodeContainerLevel)},
+			Zones: topologyv1alpha1.ZoneList{
+				{
+					Name: "node-0",
+					Type: "Node",
+					Resources: topologyv1alpha1.ResourceInfoList{
+						MakeTopologyResInfo(cpu, "32", "30"),
+						MakeTopologyResInfo(memory, "64Gi", "60Gi"),
+						MakeTopologyResInfo(nicResourceName, "16", "16"),
+					},
+				},
+				{
+					Name: "node-1",
+					Type: "Node",
+					Resources: topologyv1alpha1.ResourceInfoList{
+						MakeTopologyResInfo(cpu, "32", "30"),
+						MakeTopologyResInfo(memory, "64Gi", "60Gi"),
+						MakeTopologyResInfo(nicResourceName, "16", "16"),
+					},
+				},
+			},
+		},
+	}
```
This is a good point, but moving the NRT defs too far from test code and reusing the same defs all over again has its own drawback. I think I can start factoring out the defs used in these test in a single place and we can iterate over there.

swatisehgal commented on test/integration/noderesourcetopology_cache_test.go:
```diff
@@ -0,0 +1,449 @@
+						string(corev1.ResourceCPU):    "16",
+						string(corev1.ResourceMemory): "24Gi",
+					},
+					expectedNode: "fake-node-cache-1",
```
Seems a bit odd to specify `expectedNode` when this pod is being deleted. Would probably better to check for err here instead?

fromanirh replied:
good point. Let me rearrange the code a bit

swatisehgal commented on test/integration/noderesourcetopology_cache_test.go:
```diff
@@ -0,0 +1,449 @@
+
+func TestTopologyCachePluginWithoutUpdates(t *testing.T) {
+	verbose := "0"
+	if val, ok := os.LookupEnv("SCHED_PLUGINS_TEST_VERBOSE"); ok {
+		verbose = val
+	}
+
+	cacheResyncPeriod := int64(defaultCacheResyncPeriodSeconds)
+	if val, ok := os.LookupEnv("SCHED_PLUGINS_TEST_RESYNC_PERIOD_SECS"); ok {
```
Do we provide default values for these environment variables somewhere? Perhaps, we should do that in the Makefile?

fromanirh replied:
I'm not sure. In both cases the hardcoded defaults are expected to be good enough; if we want more verbose logs (and these are really useful to debug test failures) we have the option to. The resync period tunable OTOH is probably better to be removed.

swatisehgal replied:
I don't feel too strongly about this but my only concern was that we have not made it very transparent to someone who wants to run these integration tests that these environment variables exist and can be populated for getting more verbose log or to change the cache resync period. The only way of being aware that these environment variables exist is by inspecting this integration test files. Makefile is typically where environment variables are populated and that is why suggested that.

fromanirh replied:
Fair enough, added!

swatisehgal reviewed (commented):
Left a few minor comments but the preliminary integration tests are looking good. Looking forward to seeing more extensive integration tests especially involving the cache invalidation via the resync loop.

We had offline discussion about a possible test flow for this so capturing that here:

1. Have initial NRT where fingerprint in the cache that does not match the actual state on the node
2. Resync -> node is still dirty in the cache because of podfingerprint mismatch
3. Update cache with the correct fingerprint
4. Resync -> should mark the node clean because of podfingerprint match

fromanirh commented on test/integration/noderesourcetopology_cache_test.go:
```diff
@@ -0,0 +1,449 @@
+			},
+		},
+		{
+			name: "BU pod: with devices, pessimistic cache overallocation prevents pod to be scheduled",
```
thanks, fixed

fromanirh commented:
This PR is already big and likely increasing in size, so I'm deferring further test utils refactoring and test additions (notable the integration tests for the resync loop) in to part 3.

fromanirh commented:
In the latest upload, I addressed all comments except the one referring to the architecture of the cache and which data we store and why. I think it's beneficial to explore and explain a bit more the current state, even if later on we decide to pivot, to make sure the challenges the current architecture addresses are tackled effectively.

Huang-Wei commented on pkg/noderesourcetopology/cache/overreserve.go:
```diff
@@ -0,0 +1,228 @@
+
+		klog.V(6).InfoS("nrtcache: trying to resync NodeTopology", "logID", logID, "node", nodeName, "fingerprint", pfpExpected)
+
+		err = checkPodFingerprintForNode(logID, ov.nodeIndexer, nodeName, pfpExpected)
```
(move discussion of https://github.com/kubernetes-sigs/scheduler-plugins/pull/437#discussion_r1021133275 here.)

Let me put it another way: suppose a node has pod1, pod2 scheduled, and NRT has properly get populated with the pfp signature annotation info. Now pod3 comes in and gets (over-)reserved (maybe bound or maybe not), and pod4 comes in but not fit in the node, so it triggers `Resync()`.

At this moment, the combination of pfpExpected and the pods ov.nodeIndexer can be:

| pfpExpected        | ov.nodeIndexer     | result                                                                 |
|--------------------|--------------------|------------------------------------------------------------------------|
| [pod1, pod2]       | [pod1, pod2]       | match, flush the node incorrectly<br>b/c over-reserved pod3 is flushed |
| [pod1, pod2]       | [pod1, pod2, pod3] | mismatch                                                               |
| [pod1, pod2, pod3] | [pod1, pod2, pod3] | match, flush the node correctly                                        |
| [pod1, pod2, pod3] | [pod1, pod2]       | (this might be possible if pod3 is deleted)<br>mismatch                |

So in the first row's case, pod4's next scheduling result could be incorrect b/c over-reserved pod3 has been flushed. This case can be simulated by plumbing a time.Sleep() in binding cycle. Let me know if I misunderstand any logic here.

fromanirh replied:
Thanks for raising this point. I think you are right. If we trigger a resync:
1. AFTER a pod was overreserved
2. BUT BEFORE the pod is running
we can incorrectly flush the internal cache, lose the over-reservation and incorrectly schedule.

The fix could be adding the pods we reserved to the indexer immediately (or in general tracking the overreserved pods immediately).
The indexer should handle transparently deletion, so this fix seems pretty self-contained and safe.

fromanirh replied:
Thinking more about it, we should surely track these pods, but just adding them to the nodeIndexer is possibly not the best approach. I'll explore options (e.g. separate tracking map, other options) here.

Huang-Wei commented on pkg/noderesourcetopology/cache/overreserve.go:
```diff
@@ -0,0 +1,228 @@
+type OverReserve struct {
+	lock             sync.Mutex
+	nrts             nrtStore
+	assumedResources map[string]resourceStore // nodeName -> resourceStore
```
`resourceStore` is a pod -> v1.ResourceList, but the pod info is not used in UpdateNRT(), right?

If so, maybe we can consider simplifying the structure to make it a `map[string]v1.ResourceList` so as to make UpdateNRT() from a 3-level for loop to a 2-level for loop.

fromanirh replied:
OTOH we can get rid of `ContainsPod` and simplify at least a bit

Huang-Wei reviewed (commented):
@fromanirh thanks for the explanation! Some comments below.

fromanirh commented on pkg/noderesourcetopology/cache/overreserve.go:
```diff
@@ -0,0 +1,228 @@
+	nodeIndexer   NodeIndexer
+}
+
+func NewOverReserve(lister listerv1alpha1.NodeResourceTopologyLister, indexer NodeIndexer) (*OverReserve, error) {
```
right, missed from your previouse review. Fixing.

fromanirh commented on pkg/noderesourcetopology/cache/overreserve.go:
```diff
@@ -0,0 +1,228 @@
+	return nrt
+}
+
+func (ov *OverReserve) MarkNodeDiscarded(nodeName string, pod *corev1.Pod) {
```
seems better indeed. Changing.

fromanirh commented on pkg/noderesourcetopology/cache/overreserve.go:
```diff
@@ -0,0 +1,228 @@
+	ov.FlushNodes(logID, nrtUpdates...)
+}
+
+// Flush drops all the cached information about a given node, resetting its state clean.
```
sorry, missed in the previous upload. Fixed.

fromanirh commented on pkg/noderesourcetopology/cache/overreserve.go:
```diff
@@ -0,0 +1,228 @@
+// 2. it was pessimistically overallocated, so the node is a candidate for resync
+// This function enables the caller to know the slice of nodes should be considered for resync,
+// avoiding the need to rescan the full node list.
+func (ov *OverReserve) DirtyNodeNames() []string {
```
Sure, done.

fromanirh commented on pkg/noderesourcetopology/cache/store.go:
```diff
@@ -0,0 +1,197 @@
+type counter map[string]int
+
+func (cnt counter) Incr(key string) int {
+	val := cnt[key]
+	val++
+	cnt[key] = val
+	return val
```
done, thanks!

fromanirh commented on pkg/noderesourcetopology/cache/store.go:
```diff
@@ -0,0 +1,197 @@
+}
+
+func (cnt counter) IsSet(key string) bool {
+	return cnt[key] > 0
```
done, thanks

fromanirh commented:
I'm contemplating if I can split this PR or spin off commits to make the review more manageable. Except perhaps the first commit and perhaps https://github.com/kubernetes-sigs/scheduler-plugins/pull/437/commits/42e6b045c6e720c83d1a7c63ecc40fb611151d75 nothing pops out. I'll keep looking and I'm open to suggestions.

PiotrProkop commented on pkg/noderesourcetopology/cache/store.go:
```diff
@@ -0,0 +1,217 @@
+					// even report this error, being an expected condition
+					continue
+				}
+				if zr.Available.Cmp(qty) < 0 {
```
So `nrtcache` can only be used when POD's resources can be fit on a single NUMA node? That's fair assumption for now but I think this should be documented clearly if someone in the future would want to start playing around multi NUMA PODs and `NRT` 😄

fromanirh replied:
Hi! what I want to do here is to prevent the numa zone resources to go negative. The cache per se has no dependencies on specific TM policies (= is expected to work with any TM policy, even though we test mostly with single-numa-node so some cases may have been overisighted).
The NodeResourceTopology plugin wants to support single-numa-node policy first and foremost. We may want to add/explicitely support restricted policy later on.

PiotrProkop replied:
Ok, it makes sense since `noderesourcetopology` plugins run after `NodeResourcesFit` so we know that we can fit the POD on the node.

Huang-Wei commented:
@fromanirh I think the major unresolved comment _was_ https://github.com/kubernetes-sigs/scheduler-plugins/pull/437#discussion_r1024630227 when I reviewed it last time.

Could you let me know if this PR is now ready for another round of review? (i.e., ^^ is resolved in the latest version).

fromanirh commented:
> @fromanirh I think the major unresolved comment _was_ [#437 (comment)](https://github.com/kubernetes-sigs/scheduler-plugins/pull/437#discussion_r1024630227) when I reviewed it last time.
> 
> Could you let me know if this PR is now ready for another round of review? (i.e., ^^ is resolved in the latest version).

Hi @Huang-Wei , thanks for chiming in! I addressed your comment in https://github.com/kubernetes-sigs/scheduler-plugins/pull/437/commits/3ead89acd55e6586555c338ecbbf6af2d2e09546 . This PR is ready for another round of review anytime. If you prefer a different arrangement of commits (e.g. further split) please just let me know.

Huang-Wei commented on pkg/noderesourcetopology/cache/indexer.go:
```diff
@@ -0,0 +1,236 @@
+	if nodeName == "" {
+		// we're deleting a pod which passed the reserve stage but never got to the binding stage.
+		// Should be unlikely, but can happen!
+		nodeName = nni.podUIDToCandidateNodeMap[pod.UID]
+		wasReserved = true
```
nit:

```suggestion
		nodeName, wasReserved = nni.podUIDToCandidateNodeMap[pod.UID]
```

fromanirh replied:
right, and actually more correct. Fixed!

Huang-Wei commented on pkg/noderesourcetopology/cache/overreserve.go:
```diff
@@ -0,0 +1,234 @@
+	nodeIndexer   NodeIndexer
+}
+
+func NewOverReserve(lister listerv1alpha1.NodeResourceTopologyLister, indexer NodeIndexer) (Interface, func(), error) {
```
nit: given `NewOverReserve` is just a concrete function, it's fine to return `*OverReserve` instead of `Interface`, so that we can shorten the return args as `*OverReserve, error` (the sync func can be get from `*OverReserve`).

fromanirh replied:
fixed with the very last upload!

Huang-Wei commented on pkg/noderesourcetopology/cache/store.go:
```diff
@@ -0,0 +1,217 @@
+
+// GetByNode returns a copy of the stored Node Resource Topology data for the given node,
+// or nil if no data is associated to that node.
+func (nrs *nrtStore) GetByNode(nodeName string) *topologyv1alpha1.NodeResourceTopology {
```
maybe rename `GetByNode` (e.g., `GetNodeCopy`) to let the reader know it returns a clone from the function name.

fromanirh replied:
renamed as `GetNRTCopyByNodeName`

Huang-Wei reviewed (commented):
I did another pass and the logic looks pretty neat! Just a few nits below.

fromanirh commented on pkg/noderesourcetopology/cache/store.go:
```diff
@@ -0,0 +1,217 @@
+	}
+}
+
+type counter struct {
```
I went overboard addressing https://github.com/kubernetes-sigs/scheduler-plugins/pull/437#discussion_r1022580172 - reverting the `counter changes`.

fromanirh commented on pkg/noderesourcetopology/cache/overreserve.go:
```diff
@@ -0,0 +1,234 @@
+	assumedResources map[string]*resourceStore // nodeName -> resourceStore
+	// nodeDiscarded counts how many times a node is filtered out. This is used as trigger condition to try
+	// to resync nodes. See The documentation of Resync() below for more details.
+	nodeDiscarded *counter
```
this was me overreacting to https://github.com/kubernetes-sigs/scheduler-plugins/pull/437#discussion_r1022580172. Addressing.

fromanirh replied:
also, `nodesMaybeOverreserved` looks more consistent and a bit more expressive. Changing.

fromanirh commented on pkg/noderesourcetopology/cache/store.go:
```diff
@@ -0,0 +1,217 @@
+
+// resourceStore maps the resource requested by pod by pod namespaed name. It is not thread safe and needs to be protected by a lock.
+type resourceStore struct {
+	data map[string]corev1.ResourceList
```
thanks, added

fromanirh commented:
Addressed all comments. Thanks for the comprehensive reviews so far!

Huang-Wei commented:
Wonderful. Let's ship it 🚢 

@fromanirh really appreciate your continuous efforts in polishing the design. Especially this one that, I know this is pretty challenging, and is a new area that regular scheduler plugins haven't quite touched before. I believe this Reservation impl. can be inspiring for other plugins. (also bear with me on the slow reviewing pace, I should have responded more promptly..)

Also sincere thanks to other reviewers, @swatisehgal , @Tal-or and @PiotrProkop !

/lgtm
/approve

fromanirh commented:
> Wonderful. Let's ship it ship
> 
> @fromanirh really appreciate your continuous efforts in polishing the design. Especially this one that, I know this is pretty challenging, and is a new area that regular scheduler plugins haven't quite touched before. I believe this Reservation impl. can be inspiring for other plugins. (also bear with me on the slow reviewing pace, I should have responded more promptly..)

Thank you @Huang-Wei for the thorough review over the months (!!) which definitely lead to better design overall and better code. I'm grateful for your time spent here. I'm very aware we are navigating uncharted waters here, addressing a very specific and unique needs and constraints, and I do hope some of the novelties experimented here can be useful for the ecosystem later on. For the record, I did some research here to see if any existing approach explored in literature can be adapted to this use case, but with no luck. I'm still on the lookout and I'll be happy to learn and adjust as needed.

> Also sincere thanks to other reviewers, @swatisehgal , @Tal-or and @PiotrProkop !

Absolutely! Their help was pivotal to get this merged. Many thanks!
